batchupload.php

```diff
@@ -0,0 +1,322 @@
+}
+
+require_login($courseid, false);
+
+// Check if the setting is on.
+if (empty(get_config('block_opencast', 'batchuploadenabled_' . $ocinstanceid))) {
+    throw new moodle_exception('batchupload_errornotenabled', 'block_opencast', $redirecturl);
+}
+
+// Use block context for this page to ignore course file upload limit.
+$pagecontext = upload_helper::get_opencast_upload_context($courseid);
+$PAGE->set_context($pagecontext);
+
+$PAGE->set_pagelayout('incourse');
+$PAGE->set_title(get_string('batchupload', 'block_opencast'));
+$PAGE->set_heading(get_string('pluginname', 'block_opencast'));
+$PAGE->navbar->add(get_string('pluginname', 'block_opencast'), $redirecturl);
+$PAGE->navbar->add(get_string('batchupload', 'block_opencast'), $baseurl);
+
+// Capability check.
+if ($courseid == $SITE->id) {
+    // If upload initiated from overview page, check that capability is given in specific course or ownership.
+    if (!$series) {
+        redirect(new moodle_url('/blocks/opencast/overview_videos.php', ['ocinstanceid' => $ocinstanceid, 'series' => null]),
+            get_string('addvideonotallowed', 'block_opencast'), null,
+            notification::NOTIFY_ERROR);
+    }
+
+    $records = $DB->get_records('tool_opencast_series', ['series' => $series, 'ocinstanceid' => $ocinstanceid]);
+    $haspermission = false;
+    foreach ($records as $record) {
+        $coursecontext = context_course::instance($record->courseid, IGNORE_MISSING);
+        if ($coursecontext && has_capability('block/opencast:addvideo', $coursecontext)) {
+            $haspermission = true;
+            break;
+        }
+    }
+
+    if (!$haspermission) {
+        // Check if series is owned by user.
+        $apibridge = apibridge::get_instance($ocinstanceid);
+        $ocseries = $apibridge->get_series_by_identifier($series);
+        if (!$ocseries) {
+            redirect(new moodle_url('/blocks/opencast/overview_videos.php', ['ocinstanceid' => $ocinstanceid,
+                'series' => $series, ]),
+                get_string('connection_failure', 'block_opencast'), null,
+                notification::NOTIFY_ERROR);
+        }
+
+        if (!$apibridge->is_owner($ocseries->acl, $USER->id, $SITE->id)) {
+            redirect(new moodle_url('/blocks/opencast/overview_videos.php', ['ocinstanceid' => $ocinstanceid,
+                'series' => $series, ]),
+                get_string('addvideonotallowed', 'block_opencast'), null,
+                notification::NOTIFY_ERROR);
+        }
+    }
+} else {
+    $coursecontext = context_course::instance($courseid);
+    require_capability('block/opencast:addvideo', $coursecontext);
+
+    // Make sure apibridge is available to use.
+    if (!isset($apibridge)) {
+        $apibridge = apibridge::get_instance($ocinstanceid);
+    }
+
+    // Double check if the series is related to the course, which never happens, but we check anyway.
```

Comment:
This security fix was very important to me as well. It's not certain that only "real" teachers have the permission, for example students who were tutors once might have it as well and every computer science student is probably to some degree a 'hacker'/a bit curious what they can do in a system (One doesn't have to be especially malicious to look through this code and test out if that means you can really upload videos to any series. Maybe they would even report that issue afterwards). And even if students should never have this capability, I really wouldn't be comfortable leaving it as it was, because someone could gain that capability by fault, and in general, we don't want to rely on users (even more privileged ones) to behave nicely.

Things like these "never happen", until they do happen. :) And we want a robust plugin where we are certain that it can't possibly happen.

You could of course only accept the `series`-param if `$courseid == $SITE->id`, if you'd like that better than checking whether the series is connected to the course.

Reply:
Thanks you for your input,

> You could of course only accept the `series`-param if `$courseid == $SITE->id`, if you'd like that better than checking whether the series is connected to the course.

That check is mandatory, because we are using `$series` also when the are getting back from the `batchupload_form` (submitting the form via add videos button), therefore, if we only accept `$series` in case `$courseid == $SITE->id,` then there is no point of having to check that course-series relation at all!

Reply:
You're right, although the `$series` we get back from the `batchupload_form` wouldn't need to be checked again, since the mform already makes sure that only the options we pass in can be selected (using server-side validation). If i see it correctly, there is no usecase for redirecting the user to the upload form with a non-$SITE->id course and a $series, so if we rename the url param (for example from series to seriesid), to avoid getting the data from the mform in the `optional_param()`, we could only accept the seriesid url-param if `$courseid == $SITE->id` and otherwise drop the check whether the course is connected to the series.

Of course it also works as it is now, and you don't need to change anything, but you seemed to be unhappy with the current solution, so I just wanted to mention it :)